File: backend/app/schemas/catalog.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class PlantIn(BaseModel):
    name: str = Field(max_length=100)
    scientific_name: str | None = None
    description: str | None = None
    care_info: str | None = None
    image_url: str | None = None


class PlantUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    scientific_name: str | None = None
    description: str | None = None
    care_info: str | None = None
    image_url: str | None = None


class PlantOut(ORMModel):
    id: uuid.UUID
    name: str
    scientific_name: str | None = None
    description: str | None = None
    care_info: str | None = None
    image_url: str | None = None
    created_at: datetime | None = None


RISK = r"^(low|medium|high)$"


class DiseaseIn(BaseModel):
    plant_id: uuid.UUID | None = None
    ai_label: str | None = Field(default=None, max_length=150)
    name: str = Field(max_length=150)
    description: str | None = None
    symptoms: str | None = None
    causes: str | None = None
    treatment: str | None = None
    prevention: str | None = None
    risk_level: str | None = Field(default=None, pattern=RISK)
    image_url: str | None = None


class DiseaseUpdate(BaseModel):
    plant_id: uuid.UUID | None = None
    ai_label: str | None = None
    name: str | None = None
    description: str | None = None
    symptoms: str | None = None
    causes: str | None = None
    treatment: str | None = None
    prevention: str | None = None
    risk_level: str | None = Field(default=None, pattern=RISK)
    image_url: str | None = None


class MedicineIn(BaseModel):
    name: str = Field(max_length=150)
    active_ingredient: str | None = None
    description: str | None = None
    usage: str | None = None
    precautions: str | None = None
    manufacturer: str | None = None
    image_url: str | None = None


class MedicineOut(ORMModel):
    id: uuid.UUID
    name: str
    active_ingredient: str | None = None
    description: str | None = None
    usage: str | None = None
    precautions: str | None = None
    manufacturer: str | None = None
    image_url: str | None = None


class MedicineRecommendation(MedicineOut):
    recommendation: str | None = None


class DiseaseOut(ORMModel):
    id: uuid.UUID
    plant_id: uuid.UUID | None = None
    plant_name: str | None = None
    ai_label: str | None = None
    name: str
    description: str | None = None
    symptoms: str | None = None
    causes: str | None = None
    treatment: str | None = None
    prevention: str | None = None
    risk_level: str | None = None
    image_url: str | None = None


class DiseaseDetail(DiseaseOut):
    medicines: list[MedicineRecommendation] = []


class DiseaseMedicineLink(BaseModel):
    medicine_id: uuid.UUID
    recommendation: str | None = None


class DiseaseMedicinesIn(BaseModel):
    items: list[DiseaseMedicineLink]

File: mobile/lib/domain/entities/catalog.dart
class Plant {
  Plant({required this.id, required this.name, this.scientificName, this.description, this.careInfo, this.imageUrl});
  final String id, name;
  final String? scientificName, description, careInfo, imageUrl;

  factory Plant.fromJson(Map<String, dynamic> j) => Plant(
        id: j['id'],
        name: j['name'],
        scientificName: j['scientific_name'],
        description: j['description'],
        careInfo: j['care_info'],
        imageUrl: j['image_url'],
      );

  Map<String, dynamic> toJson() => {
        'name': name,
        'scientific_name': scientificName,
        'description': description,
        'care_info': careInfo,
      };
}

class Medicine {
  Medicine({
    required this.id,
    required this.name,
    this.activeIngredient,
    this.description,
    this.usage,
    this.precautions,
    this.manufacturer,
    this.recommendation,
  });
  final String id, name;
  final String? activeIngredient, description, usage, precautions, manufacturer, recommendation;

  factory Medicine.fromJson(Map<String, dynamic> j) => Medicine(
        id: j['id'],
        name: j['name'],
        activeIngredient: j['active_ingredient'],
        description: j['description'],
        usage: j['usage'],
        precautions: j['precautions'],
        manufacturer: j['manufacturer'],
        recommendation: j['recommendation'],
      );
}

class Disease {
  Disease({
    required this.id,
    required this.name,
    this.plantId,
    this.plantName,
    this.aiLabel,
    this.description,
    this.symptoms,
    this.causes,
    this.treatment,
    this.prevention,
    this.riskLevel,
    this.medicines = const [],
  });
  final String id, name;
  final String? plantId, plantName, aiLabel, description, symptoms, causes, treatment, prevention, riskLevel;
  final List<Medicine> medicines;

  factory Disease.fromJson(Map<String, dynamic> j) => Disease(
        id: j['id'],
        name: j['name'],
        plantId: j['plant_id'],
        plantName: j['plant_name'],
        aiLabel: j['ai_label'],
        description: j['description'],
        symptoms: j['symptoms'],
        causes: j['causes'],
        treatment: j['treatment'],
        prevention: j['prevention'],
        riskLevel: j['risk_level'],
        medicines: [for (final m in (j['medicines'] as List? ?? [])) Medicine.fromJson(m)],
      );
}
